### ISSUE TITLE
Support UDDSketch (aggregation)

### ISSUE THREAD

**Issue by Sytten** [timescale/timescaledb-toolkit#41]
## What's the functionality you would like to add ##
Also discussed here: https://github.com/timescale/timescale-analytics/discussions/35#discussioncomment-299930
I would like to strongly push for the use of UDDSketch which build upon the DDSketch from Datadog. DDSketch is great for unbounded series and UDDSketch allows for bounded series while keeping a fixed relative error (increasing with each downsampling obviously).
**DDSketch:**
- paper: https://arxiv.org/abs/1908.10693
- implementation: https://github.com/DataDog/sketches-go, (reworked for int here https://github.com/GraphMetrics/ddsketch-go)

**UDDSketch:**
- paper: https://arxiv.org/abs/2004.08604 and https://arxiv.org/abs/2101.06758
- implementation: https://github.com/cafaro/UDDSketch and https://github.com/cafaro/PUDDSKETCH

## How would the function be used ##
It is very common for metrics to be aggregated by the various agents per interval and then sent to the analytics/monitoring service. But usually we want to aggregate those histograms because we either don't need the precision or we want to create materialized views (like the P99 of the last hour per endpoint). So creating merge functions for histograms (in this case UDDSketch) is essential.

## Why should this feature be added?  ##
Currently it is possible to do everything in SQL since the actual underlining algorithm isn't that big, but it is annoying.

### What scale is this useful at? ###
All sizes IMO, you rarely want to send all data points to the server so you usually send one form of histogram or the other. In this case I advocate for the support of UDDsketch.

## Drawbacks ##
Maybe it would better to have a generic merge function that supports multiple algorithms?

## Open Questions ##
How to the deal with the input format that might be different (at graphmetrics.io we store the key, count and number of merges, but this is just one way to store the buckets).

## Alternatives ##
I created an analysis of the different algorithms we considered for graphmetrics.io here: https://docs.google.com/document/d/1CR2w1E799Ar5_3OyKowP3fsOgFAtUdsd-rJrVKArWbg/edit?usp=sharing
I still think that UDDsketch is the best for bounded histograms with a proven relative error.

**Comment by JLockerman** [comment]
see also https://github.com/timescale/timescale-analytics/issues/1

**Comment by JLockerman** [comment]
Am experimental version of this is available in [0.1.0](https://github.com/timescale/timescale-analytics/releases/tag/0.1.0), is available on Forge soon.

**Comment by Sytten** [comment]
Nice, would be great if we could have methods that works on sketches too since you generally don't store the raw data.
But it is a good first step!

**Comment by JLockerman** [comment]
@Sytten if you mean an aggregate over UDDSketches that outputs the combined UDDSketche, that's actually available in 0.1.0 (though it seems to be missing from the documentation cc @WireBaron). That said, I'd be remiss if I didn't point out that with [TimescaleDB data retention](https://docs.timescale.com/latest/using-timescaledb/data-retention), a [continuous aggregate](https://docs.timescale.com/latest/using-timescaledb/continuous-aggregates) the outputs UDDSketches will only store the sketch, and not the underlying data.

**Comment by Sytten** [comment]
Not sure I understood the last part right, but I will circle back once I read the doc. Currently I store the indexes and counts in a table:
<img width="486" alt="Screen Shot 2021-04-07 at 5 38 40 PM" src="https://user-images.githubusercontent.com/2366731/113938108-22ead200-97c8-11eb-8249-885e90bc4300.png">
So it would be nice to be able to load that so I can compute a P99 directly in the DB.

**Comment by WireBaron** [comment]
@Sytten, if I'm understanding this correctly you have a DB with indices and associated counts, and you'd like to find the index associated with the 99 percentile of the count column?

If my understanding is correct we'll probably have to add a bit more functionality to help with this, as our current UddSketch implementation will only find a column's percentile value.  That being said, making this work with timestamp, value pairs (or even index, values) is an obvious improvement we should be doing.

**Comment by Sytten** [comment]
Yes exactly, this use case is quite important since you might want to keep aggregates longer than the original data (say you do per hour P99 for up to 1 month and then per day) or you don't have the original data (my case for monitoring).

**Comment by WireBaron** [comment]
Sorry, I think I misunderstood a bit.  Your database is here is directly storing the buckets of a UddSketch you've already computed?  That's definitely a bit different from our intended case of building up the sketch directly from raw data, but we can still make this work by writing a function to convert your DB into one of our UddSketch objects.  You can then run any percentile queries on this object and combine it with other sketches if you like.  It should also be far more space efficient than the database.

To build our UddSketch, we'll need to know the original alpha of your sketch, what it's maximum number of buckets are, and how many times it's gone through key collapsing.  I also need to know how you're mapping your indexes to real numbers (i.e. is this just approximately gamma^index, or is there mapping to handle negative values).

**Comment by Sytten** [comment]
Yes I store directly DDSketch buckets since my client don't send the full data, it's a monitoring solution.
I am currently using DDSketch without collapsing and alpha of 0.01. I using the log version for the mapping and I only use positive values, I forked the datadog: https://github.com/GraphMetrics/sketches-go
if it doesn't work it's fine, but since we should be using the same formulas it should work :)

**Comment by JLockerman** [comment]
@WireBaron do you think something like this could work?

```SQL
SELECT
    format(
        '{"version":1,"alpha":0.01,"max_buckets":300,"num_buckets":%s,"compactions":0,"count":%s,"sum":0.0,"keys":[%s],"counts":[%s]}',
        num_buckets,
        total_count,
        keys,
        counts
    )::timescale_analytics_experimental.uddsketch sketch
FROM (SELECT
        count(index) num_buckets,
        sum(count) total_count,
        string_agg(format('{"Positive":%s}', index), ',') keys,
        string_agg(format('%s', index), ',') counts
    FROM ddsketch) dd
```

it creates a textual version of the UDDSketch from the inputs. I had ignore `sum` because I'm not sure if @Sytten has the information to compute that, and I picked an arbitrary `max_buckets` because I didn't know the real value. In my tests it _seems_ to work...

**Comment by WireBaron** [comment]
That actually looks much nicer than the conversion function I came up with.  It should be correct as long as the corresponding go DDSketch is using a mapping with a 0 normalizedIndexOffset (this should be the case for a `LogUnboundedDenseDDSketch` which sounds like what they're using), but there may be an off by 1 error in the translation still (@Sytten, if the results seem off by about 2% in all cases this is probably the cause).

The lack of a sum does mean that our `mean` and `sum` functions won't work, but all the quantile calculations should still be correct.

Also, one note about `max_buckets`, this should be fine to set to an arbitrary large value, but it must be >= `num_buckets`, and if you want to later combine digests, the `max_buckets` on them must match.

**Comment by WireBaron** [comment]
I did notice a couple of small errors in the snippet above, here is the corrected version
```SQL
SELECT
    format(
        '{"version":1,"alpha":0.01,"max_buckets":300,"num_buckets":%s,"compactions":0,"count":%s,"sum":0.0,"keys":[%s],"counts":[%s]}',
        num_buckets,
        total_count,
        keys,
        counts
    )::timescale_analytics_experimental.uddsketch sketch
FROM (SELECT
        count(index) num_buckets,
        sum(count) total_count,
        string_agg(format('{"Positive":%s}', index), ',') keys,
        string_agg(format('%s', count), ',') counts
    FROM ddsketch ORDER BY index) dd
```

**Comment by Sytten** [comment]
Nice I will give it a shot and compare with the results I have when computing application side!
What does sum represent in your implementation? I don't really need it anyway but curious.

**Comment by WireBaron** [comment]
Sum is the total of all the individual values.  It's really only used to compute the mean.

**Comment by JLockerman** [comment]
This was released in 1.0.0 with some changes to the text format, will do a followup on that.

**Comment by davidkohn88** [comment]
Seems like this was released and stabilized so I'm going to close.
